py/sight/widgets/decision/shared_batch_messages.py

```diff
@@ -0,0 +1,59 @@
+from dataclasses import dataclass
+from dataclasses import field
+from typing import Any, Dict, Optional
+
+
+@dataclass
+class DecisionMessage:
+  action_params: dict
+  action_id: int
+  reward: Optional[int] = None
+  discount: Optional[int] = None
+  outcome_params: Optional[dict] = None
+
+
+class CachedBatchMessages:
+  _instance = None  # Class-level variable for the singleton instance
```

Comment:
To allow for this class to be used from multiple threads or multiple Sight objects it would be more modular to keep explicit references to CachedBatchMessages objects.

Reply:
Ahh , Great idea for future cases , The following class sounds good ?

```python
class CachedBatchMessages:
  """A singleton class for caching batch messages.

  Supports multiple singleton instances distinguished by sight_id and thread_id.

  Attributes:
      batch_messages: A dictionary of cached DecisionMessages.
      _instances: A class-level dictionary to store singleton instances.
      _class_lock: A class-level lock for thread-safe instance creation.
  """

  # class vars
  _instances: ClassVar[Dict[str, 'CachedBatchMessages']] = {}
  _class_lock: ClassVar[Lock] = Lock()

  # instance vars
  _lock: threading.Lock
  _sight_id: str
  _thread_id: Optional[str]
  _unique_id: Optional[str]
  # Explicitly declare the type of batch_messages
  batch_messages: Dict[int, DecisionMessage]

  def __new__(
      cls,
      sight_id: str = 'default',
      thread_id: Optional[str] = None,
      unique_id: Optional[str] = None,
  ):
    """Create or retrieve a singleton instance for a specific sight_id and thread_id.

    Args:
        sight_id: A unique identifier for the sight.
        thread_id: An optional identifier for the thread.
        unique_id: An optional unique identifier for the instance.

    Returns:
        The singleton instance for the given configuration.
    """
    with cls._class_lock:
      # Create a unique key combining sight_id and thread_id
      instance_key = cls._create_instance_key(sight_id, thread_id, unique_id)

      # If no instance exists for this key, create a new one
      if instance_key not in cls._instances:
        instance = super().__new__(cls)
        # Initialize the instance's attributes
        instance.batch_messages: Dict[int, DecisionMessage] = {}
        instance._lock = Lock()
        instance._sight_id = sight_id
        instance._thread_id = thread_id
        instance._unique_id = unique_id

        # Store the instance in the class-level dictionary
        cls._instances[instance_key] = instance

      return cls._instances[instance_key]
```

Reply:
What I had in mind was a bit simpler. Do you think the following would work?

Keep separate instances of CachedBatchMessages that don't document their sight_id, thread_id or unique_id. Each Sight object instance would keep its own CachedBatchMessages instance, which will free us from having to include instance tracking machinery inside CachedBatchMessages as it would be implicitly done insight of Sight object instances. Further, if the Sight object needed multiple instances of CachedBatchMessages it could just create as many as needed with no changes to CachedBatchMessages.

What this means in code is
- keep all the code in all_messages() and below and not include any class static state or functions
- keep a reference to the CachedBatchMessages object inside the Sight object or even better, the SingleActionOptimizer object

Reply:
I see your point now. **Correct me if I’m wrong**—the Singleton pattern isn’t necessary here because we don’t need the `CachedBatchMessages` class to handle instance tracking itself. Instead, by keeping a `CachedBatchMessages` instance directly within the optimizer class, the optimizer effectively handles this tracking. This avoids the complexity of managing instances separately and keeps the design much simpler.
that what you meant , right ?